SDK-3807 Add custom session stores (#993)
auth0/nextjs-auth0: EXAMPLES.md

Have a look at the `basic-example` app [./examples/basic-example](./examples/basic-example).

## Create your own instance of the SDK

When you use the named exports, the SDK creates an instance of the SDK for you and configures it with the provided environment variables.

```js
// These named exports create and manage their own instance of the SDK configured with
// the provided `AUTH0_*` environment variables
import {
  handleAuth,
  handleLogin,
  handleCallback,
  handleLogout,
  handleProfile,
  withApiAuthRequired,
  withPageAuthRequired,
  getSession,
  getAccessToken
} from '@auth0/nextjs-auth0';
```

However, there are various reasons why you might want to create and manage an instance of the SDK yourself:

- You may want to create your own instance for testing

> - You may not want to use environment variables for the configuration of secrets (eg using CredStash or AWS's Key Management Service)

Reviewer: ```suggestion
- You may not want to use environment variables for the configuration of secrets (for example, to use CredStash or AWS's Key Management Service)
```

Reviewer: The doc team's guidelines instruct not to use latin abbreviations.

- You may be using a [custom session store](#use-a-custom-session-store) and need to provide the configuration as code.

In this case you can use the [initAuth0](https://auth0.github.io/nextjs-auth0/modules/instance.html) method to create an instance.

```js
// utils/auth0.js
import { initAuth0 } from '@auth0/nextjs-auth0';

export default initAuth0({
  secret: 'LONG_RANDOM_VALUE',
  issuerBaseURL: 'https://your-tenant.auth0.com',
  baseURL: 'http://localhost:3000',
  clientID: 'CLIENT_ID',
  clientSecret: 'CLIENT_SECRET'
});
```

Now rather than using the named exports, you can use the instance methods directly.

```js
// pages/api/auth/[...auth0].js
import auth0 from '../../../utils/auth0';

// Use the instance method
export default auth0.handleAuth();
```

> Note: You should not use the instance methods in combination with the named exports,
> otherwise you will be creating multiple instances of the SDK. For example:

```js
// DON'T Mix instance methods and named exports
import auth0 from '../../../utils/auth0';
import { handleLogin } from '@auth0/nextjs-auth0';

> export default auth0.handleAuth({
>   // <= instance method

Reviewer: ```suggestion
export default auth0.handleAuth({ // <= instance method
```

  return withPageAuthRequired({ returnTo })(ctx);
};
```

## Use a custom session store

You need to create your own instance of the SDK in code, so you can pass an instance of your session store to the SDK's configuration.

```typescript
// lib/auth0.ts
import { SessionStore, SessionStorePayload, initAuth0 } from '@auth0/nextjs-auth0';

class Store implements SessionStore {
  private store: KeyValueStoreLikeRedis<SessionStorePayload>;
  constructor() {
    // If you set the expiry accross the whole store use the session config,

>     // e.g. `min(config.session.rollingDuration, config.session.absoluteDuration)`

Reviewer: ```suggestion
    // for example `min(config.session.rollingDuration, config.session.absoluteDuration)`
```

Reviewer: The doc team's guidelines instruct not to use latin abbreviations.

>     // the default is 24hrs

Reviewer: ```suggestion
    // the default is 24 hrs
```

    this.store = new KeyValueStoreLikeRedis();
  }
  async get(id) {
    const val = await this.store.get(id);
    return val;
  }
  async set(id, val) {
    // To set the expiry per item, use `val.header.exp` (in secs)

>     const expiryMs = val.header.exp * 1000;

Reviewer: Where is `expiryMs` used?

Author: Fair, point - it depends on the store, will add an example for redis

```suggestion
    const expiryMs = val.header.exp * 1000;
    // Example for Redis: redis.set(id, val, { pxat: expiryMs });
```
